asllib/ASLTypingReference.mld

```diff
@@ -16,107 +16,753 @@ welcome; those can be sent to Arm’s Architecture Formal Team Lead Jade
+    - ty is S.
+
+  - All of the following applies:
+    - S and T are both named types;
+    - ty is a common supertype of S and T;
+    - ty is a subtype of all other common supertypes of S and T.
+
+  - All of the following applies:
+    - S and T both have the structure of array types with the same index type
+      and the same element types;
+    - One of the following applies:
+      - All of the following applies:
+        - S is a named type;
+        - T is an anonymous type;
+        - ty is S.
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - ty is T.
+
+  - All of the following applies:
+    - S and T both have the structure of tuple types with the same number of elements;
+    - The types of the elements of S type-satisfy the types of the elements of T;
+    - The types of the elements of T type-satisfy the types of the elements of S;
+    - One of the following applies:
+
+      - All of the following applies:
+        - S is a named type;
+        - T is an anonymous type;
+        - ty is S.
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - ty is T.
+
+     - All of the following applies:
+        - S is an anonymous type;
+        - T is an anonymous type;
+	- ty is the tuple type where the type of each element is the lowest common
+	  ancestor of the types of the corresponding elements of S and T. 
+
+  - All of the following applies:
+    - S and T both have the structure of well-constrained integer types;
+    - One of the following applies:
+
+      - All of the following applies:
+        - S is a named type;
+        - T is an anonymous type;
+        - ty is S.
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - ty is T.
+
+     - All of the following applies:
+        - S is an anonymous type;
+        - T is an anonymous type;
+	- ty is the well-constrained integer type whose domain is the union of the
+	  domains of S and T.      
+
+  - All of the following applies:
+    - Either S or T have the structure of an unconstrained integer type;
+    - One of the following applies:
+
+      - All of the following applies:
+        - S is a named type;
+        - S has the structure of an unconstrained integer type;
+        - T is an anonymous type;
+        - ty is S.
+
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - T has the structure of an unconstrained integer type;
+        - ty is T.
+
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is an anonymous type;
+	- ty is the unconstrained integer type. 
+
+  - All of the following applies:
+    - Either S or T have the structure of an under-constrained integer type;
+    - ty is the under-constrained integer type. 
+
+  - ty is undefined.
```

Comment:
I think that this is dangerous in a specification, and that we should just remove this line.

Maybe we could add a comment saying that the lowest common ancestor is a partial function on pairs of types.